File: src/devices/tca9548a.cpp
#include "devices/tca9548a.h"

TCA9548A::TCA9548A(int dev, int bus, int address) : I2C(dev,bus,address){
     this->current_channel = 0;
}

TCA9548A::~TCA9548A(){}

int TCA9548A::get_selected_channel(){ return this->current_channel; }

int TCA9548A::select_channel(uint8_t channel, bool verbose, bool debug){
     if(channel > 7) return -1;
     uint8_t value = (0x01 << channel);
     if(verbose) printf("[INFO] TCA9548A::select_channel() ---- Writing Value 0x%02X to i2c mux\r\n",value);
     int err = this->write_raw_byte(value,debug);
     if(err < 0) return -1;
     else{
          this->current_channel = channel;
          return 0;
     }
}

void TCA9548A::scan_bus(I2C* targetDev, bool verbose){
     int err;
     for(int ch = 0; ch < 8; ch++){
          this->select_channel(ch);
          for(int ad = 0; ad < 128; ad++){
               if(ad == this->_address) continue;
               // err = this->write_byte(ad,0);
               err = targetDev->write_raw_byte(0);
               if(err < 0){
                    // printf("[ERROR] TCA9548A::scan_bus() ---- Could not write to address [0x%02X] on channel \'%d\' with error code = %d\r\n", ad, ch, err);
               }else{
                    if(ad == targetDev->get_address())
                         printf("[INFO] TCA9548A::scan_bus() ---- Found Target device [0x%02X] at address 0x%02X on Channel %d\r\n",targetDev->get_address(), this->_address, ch);
               }
          }
     }
}

File: scripts/devices/tca9548a.py
from i2c_device import i2cDevice

class TCA9548A(object):
    def __init__(self, address=0x70, busnum=1, i2c_interface=None):
        self.address = address
        self.busnum = busnum
        self.i2c = i2cDevice(address=address, busnum=busnum, i2c_interface=i2c_interface)

    def __del__(self):
        print("Killing TCA9548A...")

    def select_channel(self,channel,address=None):
        if(address is None): address = self.address
        try:
            self.i2c.bus.write_byte(address,2**channel)
            return 1
        except: return 0

    def scan_bus(self, target, address=None, delay=None, verbose=False, debug=False):
        if(address is None): address = self.address
        # Search through all 8 channels
        for ch in range(8):
            self.select_channel(address, ch)
            # Search through all supported address id's
            for i in range(128):
                if(i == address): continue
                data = 0
                try:
                    self.i2c.bus.write_byte(i,0)
                    if(i == target): print("Found Target [0x%02X] at address 0x%02X on Channel %d"% (i,address,ch))
                    else:
                        if(verbose): print("Found device [0x%02X] at address 0x%02X on Channel %d" % (i,address,ch))
                except:
                    if(debug): print("Could not write to address %d on Bus %d" % (i,ch))
                    pass
                if(delay is not None): time.sleep(delay)


if __name__ == "__main__" :
    mux = TCA9548A()
    mux.scan_bus(0x28)
    print("---------------")
